server/src/main/java/org/elasticsearch/search/aggregations/timeseries/InternalTimeSeries.java

```diff
@@ -186,38 +194,46 @@ protected void doWriteTo(StreamOutput out) throws IOException {
 
     @Override
     public InternalAggregation reduce(List<InternalAggregation> aggregations, AggregationReduceContext reduceContext) {
-        // We still need to reduce in case we got the same time series in 2 different indices, but we should be able to optimize
-        // that in the future
-        Map<Map<String, Object>, List<InternalBucket>> bucketsList = null;
-        for (InternalAggregation aggregation : aggregations) {
-            InternalTimeSeries timeSeries = (InternalTimeSeries) aggregation;
-            if (bucketsList != null) {
-                for (InternalBucket bucket : timeSeries.buckets) {
-                    bucketsList.compute(bucket.key, (map, list) -> {
-                        if (list == null) {
-                            list = new ArrayList<>();
+        // TODO: optimize single result case either by having a if check here and return aggregations.get(0) or
+        // by overwriting the mustReduceOnSingleInternalAgg() method
+
+        final BigArrays bigArrays = reduceContext.bigArrays();
+        final int initialCapacity = aggregations.stream()
+            .map(value -> (InternalTimeSeries) value)
+            .mapToInt(value -> value.getBuckets().size())
+            .max()
+            .getAsInt();
+        try (LongObjectPagedHashMap<List<InternalBucket>> tsKeyToBuckets = new LongObjectPagedHashMap<>(initialCapacity, bigArrays)) {
```

Comment:
Do we send these back in sorted order? Should we? That way we don't need to do this collection thing - we can just walk the results in parallel and marge.

Reply:
Yes, I think the result is sorted. We collect the tsids in order in `BytesKeyedBucketOrds` and then we emit the tsids from `BytesKeyedBucketOrds` in the same way it was inserted.

Reply:
I think we probably should have a stronger guarantee - that the results are sorted by the bytes representation of the tsid. I think they already *are*. But we should, like, force it. Or at least `assert` it.

Reply:
I've added asserts for this: 0d20460c6a4d7b3bab9d288add6a80fff2343d73